### PR TITLE
Adding configuration and API of KVStore

### PR DESCRIPTION
**Pull request by dannybenor**

### CODE REVIEW

**Comment by dannybenor** [comment]
@AnotherButler for your review

**Comment by MarceloSalazar** [comment]
The doc is looking good!
There are many new concepts. I'd like to see an example app on how KVStore can be used to help understand some of these concepts. Thanks!

File: docs/api/storage/storage.md [hunk 1]
```diff
@@ -2,9 +2,38 @@
 
 The storage APIs present in Arm Mbed OS are:
 
+- [KVStore:](#Components-implementing-KVStore-API) a common interface for components presenting set/get API.
```
[the remark this answers is not in the document]
Comment (dannybenor) [reply docs/api/storage/storage.md]: It is a WIP. Amanda is writing a technology guide and we need to add many links

File: docs/api/storage/KVStoreGlobalAPI.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+#### `full_prefix`
+
+This parameter is used in the iterator `kv_iterator_open` function to define the subset of file names that `kv_iterator_next` returns. The string is used as a prefix, and `kv_iterator_next` returns each name that starts with the exact string passed in `full_prefix`.
+
+<span class="notes">**Note:** The full_prefix string must include the `/path/`. For example: `/kv/`.</span>
+
+### Static Global API class reference
+
+<insert link>
```
Comment (AnotherButler) [inline docs/api/storage/KVStoreGlobalAPI.md]: We can't add this until the code is on master.

File: docs/api/storage/KVStoreGlobalAPI.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ 
+### Static Global API example
+
+<insert link>
```
Comment (AnotherButler) [inline docs/api/storage/KVStoreGlobalAPI.md]: Please let me know when an example exists for this API.

File: docs/api/storage/KVStoreGlobalAPI.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+### Related content
+
+<insert links>
```
Comment (AnotherButler) [inline docs/api/storage/KVStoreGlobalAPI.md]: I was planning on adding links to the other pages in this PR. Are there any others I should add?

File: docs/api/storage/KVStoreGlobalAPI.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+- **KV_REQUIRE_CONFIDENTIALITY_FLAG:** The system encrypts the data using an AES CTR, a random IV and a key derived from DeviceKey. This flag will be ignored if you select the `TDB_INTERNAL` configuration because the internal memory is seen as protected from physical attacks.
+- **KV_REQUIRE_INTEGRITY_FLAG:** The system calculates a CMAC using a key derived from DeviceKey and checks the data validity when reading from the file (function `get`).
+- **KV_REQUIRE_REPLAY_PROTECTION_FLAG:** The system keeps a copy of the data CMAC in internal memory and checks that the data CMAC corresponds to this saved CMAC. It does this to prevent an attacker replacing the latest data with a valid old version of the data. This flag will be ignored if you select the configuration `TDB_EXTERNAL_NO_RBP` or `FILESYSTEM_NO_RBP`.
+
+#### `full_prefix`
+
+This parameter is used in the iterator `kv_iterator_open` function to define the subset of file names that `kv_iterator_next` returns. The string is used as a prefix, and `kv_iterator_next` returns each name that starts with the exact string passed in `full_prefix`.
```
Comment (AnotherButler) [inline docs/api/storage/KVStoreGlobalAPI.md]: Query: What uses this parameter in the iterator function? Could we say "The iterator `kv_iterator_open` function uses this parameter to define ..."?

Also, what uses the string as a prefix? The parameter?

File: docs/api/storage/KVStoreGlobalAPI.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+The global API uses **/path/** to identify the KVStore component that receives the call. The global API removes the **/path/** from the string passed to the KVStore API. The **path** string must be equal to the string defined for the **default_kv** parameter in the API's `mbed_lib.json` file (`/kv/` by default). Use **/kv/myfilename** when using the default configuration.
+
+<span class="notes">**Note:** The configuration passes the value of the parameter `default_kv` to the function `kv_map.attach(STR(MBED_CONF_STORAGE_DEFAULT_KV),&kvstore_config)` to create the relation between the path string and the allocated KVStore component.</span>
+
+#### Creation flags
+
+You can "or" (|) the flags below and pass them to the function set in the parameter `create_flags`. These flags indicate the security required for the data saved. You can override this security behavior by explicitly selecting a configuration in the `storage_type` parameter:
+
+- **KV_WRITE_ONCE_FLAG:** The system does not an additional call to the function set with the same file name to delete or replace data.
```
Comment (AnotherButler) [inline docs/api/storage/KVStoreGlobalAPI.md]: Note to self: Add "allow" between "does" and "not" after this round of review.

**Comment by AnotherButler** [comment]
After looking at it, let's leave the file system and block device content on the storage overview page. I don't like duplicating it, and I like having that content on that page.

File: docs/api/storage/KVStoreAPI.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
+# Components implementing KVStore API
```
Comment (AnotherButler) [inline docs/api/storage/KVStoreAPI.md]: Can you explain the difference between this API and the global here? When would someone use this one over that one?

Comment (AnotherButler) [reply docs/api/storage/KVStoreAPI.md]: Also, can you explain what we mean by components and how to use them?

Comment (dannybenor) [reply docs/api/storage/KVStoreAPI.md]: Added explanation
Components-->classes

File: docs/api/storage/KVStoreGlobalAPI.md (hunk 5 above)
Comment (dannybenor) [reply docs/api/storage/KVStoreGlobalAPI.md]: changed

**Comment by AnotherButler** [comment]
It looks like some of my changes were overwritten. Can you please fix this?

**Comment by AnotherButler** [comment]
This PR is waiting for my edits to come back.

Then, there's duplication between KVStore (not global) and storage overview page. On the overview page, I want to replace the detailed description of the KVStore classes with a bulleted list of the classes. I want to keep the KVStore page as is.

Then, we need to add Doxy links and examples to KVStore. We also need to add Doxy links and examples to KVStore global.

**Comment by dannybenor** [comment]
@AnotherButler very sorry. Don't know what happened to your changes

**Comment by AnotherButler** [comment]
Note: The file I renamed is causing some issues because of case sensitivity. I'm going to merge this and then remove the duplicated content separately.

**Pull request merged**